frcxx_nft/src/state.rs

```diff
@@ -0,0 +1,207 @@
+//! Abstraction of the on-chain state related to NFT accounting
+use cid::multihash::Code;
+use cid::Cid;
+use fvm_ipld_amt::Amt;
+use fvm_ipld_amt::Error as AmtError;
+use fvm_ipld_blockstore::Block;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::tuple::*;
+use fvm_ipld_encoding::CborStore;
+use fvm_ipld_encoding::DAG_CBOR;
+use fvm_ipld_hamt::Error as HamtError;
+use fvm_ipld_hamt::Hamt;
+use fvm_shared::ActorID;
+use thiserror::Error;
+
+pub type TokenID = u64;
+
+#[derive(Error, Debug)]
+pub enum StateError {
+    #[error("ipld amt error: {0}")]
+    IpldAmt(#[from] AmtError),
+    #[error("ipld hamt error: {0}")]
+    IpldHamt(#[from] HamtError),
+    #[error("other error: {0}")]
+    Other(String),
+}
+
+/// Each token stores its owner, approved operators etc.
```

Comment:
It might be worth profiling this vs a less normalised approach after we've done more. This scheme can reduce the number of AMT traversals when fetching all the token data, but requires loading irrelevant bytes from state when only one field of the data is needed (c.f. row vs column stores). The metadata key in this case is likely to dominate the storage costs, but how often will it be read or written or-chain?

This metadata is also probably the thing most likely to be customised by users.